guides/javascript.md

```diff
@@ -0,0 +1,264 @@
+
+🚫
+
+``` javascript
+function makeRequest() {
+  return fetch("https://google.com")
+    .then((response) => {
+      return response.json().then((json) => {
+        return json['page_views'];
+      })
+    })
+}
+```
+
+✅
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch("https://google.com");
+  const json = await response.json();
+  return json['page_views'];
+}
+```
+
+## `await` promises before returning them
+
+An `async` function that returns a rejected promise created via another `async` function may disappear from the stack trace. This is solved by `await`ing the promise before returning it.
+
+<details><summary><b>Read more</b></summary>
+<br/>
+<p>If you save the following to a file (say, <code>/tmp/example.js</code>) and run it with <code>node</code>:</p>
+
+``` javascript
+async function foo() {
+  return bar();
+}
+
+async function bar() {
+  await Promise.resolve();
+  throw new Error('BEEP BEEP');
+}
+
+foo().catch(error => console.log(error.stack));
+```
+
+then you will see the following in the terminal (as of Node 18):
+
+```
+Error: BEEP BEEP
+    at bar (/private/tmp/example.js:7:9)
+```
+
+Notice how `foo` is completely missing from the stack trace!
+
+However, if you put an `await` before the call to `bar`:
+
+``` javascript
+async function foo() {
+  return await bar();
+}
+
+async function bar() {
+  await Promise.resolve();
+  throw new Error('BEEP BEEP');
+}
+
+foo().catch(error => console.log(error.stack));
+```
+
+you will now see it at the bottom of the stack trace:
+
+```
+Error: BEEP BEEP
+    at bar (/private/tmp/example.js:7:9)
+    at async foo (/private/tmp/example.js:2:10)
+```
+
+</details>
+
+🚫
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch('https://some/url');
+  return response.json();
+}
+```
+
+✅
+
+``` javascript
+async function makeRequest() {
+  const response = await fetch('https://some/url');
+  return await response.json();
+}
+```
+
+> [!WARNING]\
+> `await`ing promises has a significant performance impact when transpiling to ES6. [**Read more**](https://arthur.place/the-cost-of-return-await).
+
+## Use Jest's mock functions instead of Sinon
+
+Jest incorporates most of the features of Sinon with a slimmer API:
+
+* `jest.fn()` can be used in place of `sinon.stub()`.
+* `jest.spyOn(object, method) can be used in place of `sinon.spy(object, method)`.
+* `jest.spyOn(object, method)` can be used in place of `sinon.stub(object, method)` (with the caveat that the method being spied upon will still be called by default).
+* `jest.useFakeTimers()` can be used in place of `sinon.useFakeTimers()` (though note that Jest's "clock" object has a different API than Sinon's and offers fewer options for advancing time).
```

Comment:
As of Jest v29.5, they might finally have all of the Sinon fake timer API methods exposed (see https://jestjs.io/docs/jest-object#jestadvancetimersbytimeasyncmstorun)

I was about to comment that maybe we shouldn't recommend Jest fake timers for now, but I guess it should be safe now. Maybe instead of this disclaimer about Jest having a different API with fewer options, we can add a note about the API missing features in versions prior to v29.5? e.g.

```suggestion
* `jest.useFakeTimers()` can be used in place of `sinon.useFakeTimers()` (though note that Jest's "clock" object had fewer features than Sinon's prior to Jest v29.5).
```